tao_compiler/mlir/disc/IR/disc_shape_ops.cc

```diff
@@ -296,6 +298,68 @@ void TieShapeOp::getCanonicalizationPatterns(RewritePatternSet& results,
 
 LogicalResult TieShapeOp::verify() { return Verify(*this); }
 
+//===----------------------------------------------------------------------===//
+// SymbolicDimOp
+//===----------------------------------------------------------------------===//
+
+void SymbolicDimOp::getCanonicalizationPatterns(RewritePatternSet& results,
+                                                MLIRContext* context) {}
+
+LogicalResult SymbolicDimOp::verify() { return Verify(*this); }
+
+int64_t SymbolicDimOp::getDimSize() {
+  if (auto attr = (*this)->getAttrOfType<IntegerAttr>("value"))
+    return attr.getInt();
+  return ShapedType::kDynamicSize;
+}
+
+void SymbolicDimOp::setDimSize(int64_t val) {
+  OpBuilder b(*this);
+  (*this)->setAttr("value", b.getI64IntegerAttr(val));
+  if (val == -1) {
+    setKnownNegativeOne(true);
+  }
+}
+
+bool SymbolicDimOp::isDynamic() {
+  return getDimSize() == ShapedType::kDynamicSize;
+}
+
+void SymbolicDimOp::setKnownNonNegative(bool flag) {
+  OpBuilder b(*this);
+  (*this)->setAttr("knownNonNegative", b.getBoolAttr(flag));
+}
+
+void SymbolicDimOp::setKnownNegativeOne(bool flag) {
+  OpBuilder b(*this);
+  (*this)->setAttr("knownNegativeOne", b.getBoolAttr(flag));
+  if (flag) (*this)->setAttr("knownNonSizeOne", b.getBoolAttr(flag));
+}
+
+void SymbolicDimOp::setKnownNonSizeOne(bool flag) {
+  OpBuilder b(*this);
+  (*this)->setAttr("knownNonSizeOne", b.getBoolAttr(flag));
+}
+
+LogicalResult SymbolicDimOp::Merge(SymbolicDimOp other) {
+  if (!isDynamic() && !other.isDynamic() && getDimSize() != other.getDimSize())
```

Comment:
So, if SymbolicDimOp A corresponding to A is a Tensor, while SymbolicDimOp corresponding to B is a vector or a scalar, it is required an explicit broadcast node in inserted before getting the SymbolicDimOp B?

Reply:
Yes. Implicit broadcast semantic is materialized when converting from `cmhlo` dialect to `mhlo` dialect.  In `mhlo` dialect, broadcast is explicit.